pytorch_lightning/profiler/pytorch.py

```diff
@@ -0,0 +1,681 @@
+from typing import Any, Callable, List, Optional
+
+import torch
+
+from pytorch_lightning import _logger as log
+from pytorch_lightning.profiler.profilers import BaseProfiler
+from pytorch_lightning.utilities import _TORCH_GREATER_EQUAL_1_8, rank_zero_only
+from pytorch_lightning.utilities.cloud_io import get_filesystem
+from pytorch_lightning.utilities.distributed import rank_zero_warn
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+if _TORCH_GREATER_EQUAL_1_8:
+    from torch.autograd.profiler import record_function
+    from torch.profiler import ProfilerAction, ProfilerActivity, tensorboard_trace_handler
+
+
+class RegisterRecordFunction:
+    """
+    While profiling autograd operations, this class will add label with module name
+    around the forward function.
+
+    The Lightning PyTorch Profiler will activate this feature automatically.
+
+    It can be desactivated as follow:
+
+    Example::
+
+        from pytorch_lightning.profilers import PyTorchProfiler
+
+        profiler = PyTorchProfiler(record_module_names=False)
+
+        Trainer(profiler=profiler)
+
+    It can be used outside of Lightning as follow:
+
+    Example::
+
+        from pytorch_lightning import Trainer, seed_everything
+
+        with RegisterRecordFunction(model):
+            out = model(batch)
+
+    """
+
+    def __init__(self, model):
+        self._model = model
+        self._records = {}
+        self.handles = {}
+
+    def _start_recording(self, module, input, module_name: str = None, is_built_in: bool = None):
+        if module_name is not None:
+            record_name = module_name if is_built_in else f"{type(module)}: {module_name}"
+            self._records[record_name] = record_function(record_name).__enter__()
+        return input
+
+    def _stop_recording(self, module, input, result, module_name: str = None, is_built_in: bool = None):
+        if module_name is not None:
+            record_name = module_name if is_built_in else f"{type(module)}: {module_name}"
+            self._records[record_name].__exit__(None, None, None)
+        return result
+
+    def __enter__(self):
+        built_in_modules = dir(torch.nn)
+        for module_name, module in self._model.named_modules():
+            is_built_in = module in built_in_modules
+            pre_handle = module.register_forward_pre_hook(
+                partial(self._start_recording, module_name=module_name, is_built_in=is_built_in)
+            )
+            post_handle = module.register_forward_hook(
+                partial(self._stop_recording, module_name=module_name, is_built_in=is_built_in)
+            )
+            self.handles[module_name] = [pre_handle, post_handle]
+
+    def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any):
+        for module_name, _ in self._model.named_modules():
+            for h in self.handles[module_name]:
+                h.remove()
+
+
+class LegacyPyTorchProfiler(BaseProfiler):
```

Comment:
this PR is quite big so it'd be nice to split out the changes to the existing pytorch profiler, marking it as a legacy one into one PR and then adding support for the 1.8 profiler API in another PR

Reply:
Jirka already opened this PR which got merged: https://github.com/PyTorchLightning/pytorch-lightning/pull/6261